### ISSUE TITLE
Error: Table 'backdrop1db.cache_token' doesn't exist on upgrade from Drupal 7

### ISSUE THREAD

**Issue by leeksoup** [backdrop/backdrop-issues#5974]
## Description of the bug

I followed the directions at the docs site to upgrade from Drupal 7 to Backdrop. Everything went fine until I got partway through the update.php step, when it stopped with the below error.

## Steps To Reproduce

To reproduce the behavior:
1. Follow all the steps in https://docs.backdropcms.org/documentation/upgrading-from-drupal
2. Run the upgrade.php script by pointing browser to: http://localhost/backdrop1/core/update.php
(This is my test/ dev environment.)

## Actual behavior

upgrade.php script stopped with the following error:

Error
Error message
Notice: Trying to access array offset on value of type bool in dblog_update_1001() (line 108 of /var/www/html/backdrop1/core/modules/dblog/dblog.install).
PDOException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'backdrop1db.cache_token' doesn't exist: TRUNCATE {cache_token} ; Array ( ) in backdrop_flush_all_caches() (line 8601 of /var/www/html/backdrop1/core/includes/common.inc).
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'backdrop1db.cache_token' doesn't exist: TRUNCATE {cache_token} ; Array ( )

## Expected behavior

update.php completes normally and I can use the upgraded site. :)

## Additional information

Add any other information that could help, such as:
- Backdrop CMS version: latest (1.24)
- Web server and its version: Apache httpd 2.4.54
- Operating System and its version: Fedora 35
- Browser(s) and their versions: Brave Version 1.46.144 Chromium: 108.0.5359.128 (Official Build) (64-bit)

**Comment by klonos** [comment]
Thanks for taking the time to report this @leeksoup 🙏🏼 ...can you please also provide php and MariaDB/MySQL versions?

**Comment by klonos** [comment]
...the following might be relevant:
- https://drupal.stackexchange.com/questions/16511/cache-token-table-doesnt-exist-after-drupal-6-to-drupal-7-upgrade
- https://www.drupal.org/project/token/issues/1512934

@leeksoup was your site D6 which was upgraded to D7? Also, which version of the Token module was installed on the D7 site? (was it updated to the latest version and has update.php been run on the D7 site?)

**Comment by argiepiano** [comment]
@leeksoup, in addition to what @klonos said above, a couple of things to check. These errors are tricky to debug, as they may be caused by a number of things, including modules that are or are not supposed to be enabled, etc.

- Is the module Token enabled in your D7 site? 
- Inspect your D7 database after your've prepared for upgrade. Does it contain the table `cache_token`?  
- Can you provide a backtrace before the the fatal error happens? If you don't use xdebug you may need to add:  `if ($bin == 'token') debug(debug_backtrace());` to line 8601 of `common.inc` right before `cache($bin)->flush();`. This will generate a huge amount of text, but it will allow you to know which function called `backdrop_flush_all_caches()` and give us clues of what's going on
- It may be a good idea for you to also post here a full list of enabled D7 modules in your D7 site right before you do the upgrade

**Comment by klonos** [comment]
@leeksoup sorry for the "barrage" of questions, but we never had anyone else report these specific issues before, and we need to be able to reproduce the problem before we can fix things - otherwise there's little we can do. Please try to provide as much of the additional information we have asked, and thanks in advance for that. Let us know if you need any help or have any questions.

**Comment by leeksoup** [comment]
@klonos and @argiepiano - Thank you for your quick responses. I will try to answer all your questions below.
MariaDB version 15.1 Distrib 10.5.18
PHP version 7.4

Was your site D6 which was upgraded to D7? Yes, several years ago. I will check that link you provided.
Also, which version of the Token module was installed on the D7 site? 7.x-1.7
Was it updated to the latest version and has update.php been run on the D7 site? It was disabled. update.php was run -- I updated everything to the latest but Token probably was not updated since it was disabled.

Is the module Token enabled in your D7 site? No, disabled
Inspect your D7 database after your've prepared for upgrade. Does it contain the table cache_token? No
Can you provide a backtrace before the the fatal error happens? If you don't use xdebug you may need to add: if ($bin == 'token') debug(debug_backtrace()); to line 8601 of common.inc right before cache($bin)->flush();. This will generate a huge amount of text, but it will allow you to know which function called backdrop_flush_all_caches() and give us clues of what's going on. OK, I will try this and report back, after I try the suggestion in @klonos's second reply.

It may be a good idea for you to also post here a full list of enabled D7 modules in your D7 site right before you do the upgrade:
 Chaos  Chaos tools (ctools)                 Module  Enabled        7.x-1.21 
 Core   Block (block)                        Module  Enabled        7.94     
 Core   Comment (comment)                    Module  Enabled        7.94     
 Core   Database logging (dblog)             Module  Enabled        7.94     
 Core   Field (field)                        Module  Enabled        7.94     
 Core   Field SQL storage                    Module  Enabled        7.94     
 Core   Field UI (field_ui)                  Module  Enabled        7.94     
 Core   File (file)                          Module  Enabled        7.94     
 Core   Filter (filter)                      Module  Enabled        7.94     
 Core   Image (image)                        Module  Enabled        7.94     
 Core   List (list)                          Module  Enabled        7.94     
 Core   Menu (menu)                          Module  Enabled        7.94     
 Core   Node (node)                          Module  Enabled        7.94     
 Core   Number (number)                      Module  Enabled        7.94     
 Core   Options (options)                    Module  Enabled        7.94     
 Core   Path (path)                          Module  Enabled        7.94     
 Core   Search (search)                      Module  Enabled        7.94     
 Core   System (system)                      Module  Enabled        7.94     
 Core   Taxonomy (taxonomy)                  Module  Enabled        7.94     
 Core   Text (text)                          Module  Enabled        7.94     
 Core   Update manager (update)              Module  Enabled        7.94     
 Core   User (user)                          Module  Enabled        7.94     
 Field  Entity Reference (entityreference)   Module  Enabled        7.x-1.6  
 Other  Entity API (entity)                  Module  Enabled        7.x-1.10 
 User   jQuery Update (jquery_update)        Module  Enabled        7.x-4.1  
 Views  Views (views)                        Module  Enabled        7.x-3.28 
 Views  Views UI (views_ui)                  Module  Enabled        7.x-3.28 
 Core   Bartik (bartik)                      Theme   Enabled        7.94     
 Core   Garland (garland)                    Theme   Enabled        7.94     
 Core   Seven (seven)                        Theme   Enabled        7.94     
 Other  Bootstrap (bootstrap)                Theme   Enabled        7.x-3.26 
 Other  <custom theme> (<themename>)        Theme   Enabled

**Comment by argiepiano** [comment]
Quick thought: try enabling Token in your D7 site, update it if needed, and then try the upgrade into B again.

**Comment by leeksoup** [comment]
@argiepiano - I had the same thought after reading those links from @klonos!

Just tried it and this time the core/update.php did complete! (So it looks like the resolution is to enable/ update the Token module to recreate that cache table.) However, i have the following error message. Is this normal?

Notice: Trying to access array offset on value of type bool in dblog_update_1001() (line 108 of /var/www/html/backdrop1/core/modules/dblog/dblog.install).

Thank you both for all the help!

**Comment by argiepiano** [comment]
Glad it worked! The notice you are getting is because (apparently) the date format called `short` did not exist or was misconfigured in your D7 (visit `/admin/config/regional/date-time` in your D7 site)

**Comment by leeksoup** [comment]
Great, thank you! I will do that.

I'm also running into a permissions problem where the config files don't get written ... but that is a separate issue. Thanks again @klonos + @argiepiano

**Comment by klonos** [comment]
> So it looks like the resolution is to enable/ update the Token module to recreate that cache table.

Either that, or fully uninstall the disabled Token module in the D7 site, which will remove the table altogether.

> The `dblog_update_1001()` function is an update function that sets the format for the database error logging. You can manually edit that format in your new Backdrop site at `/admin/config/development/logging`. That should take care of it.

Sounds like we could be doing some defensive coding in the `dblog_update_1001()` function then, to avoid these errors. I've raised #5977 as a follow-up 😉
